### ISSUE TITLE
Reduce page weight

### ISSUE THREAD

**Issue by danielfernandez** [thymeleaf/thymeleaf#11]
From: http://sourceforge.net/tracker/?func=detail&aid=3495263&group_id=509826&atid=2072662

The Mozilla "Tips for authoring fast-loading HTML pages" recommend that the HTML page is stripped of whitespace before being delivered to the client.

https://developer.mozilla.org/en/HTML/Tips_for_authoring_fast-loading_HTML_pages#Reduce_page_weight

It would be extremely useful to enable/disable an option to allow whitespace to be stripped (using a bean property?) in the template mode.

**Comment by usethe4ce** [comment]
A few thoughts on this. 
- There are a few ways to compact white space. One is to just discard whitespace-only nodes (a la `xml:space`). A more aggressive way is like `normalize-space()`, discarding leading and trailing space and collapsing contiguous white space into a single space. 
- White space from the template itself is _usually_ not important, except in certain elements. This is mostly indentation to make the template readable. Discarding excess white space early could make processing substantially more efficient, besides the reduced page weight. 
- As for dynamically generated content, via `th:text` especially, it's harder to say. Safer to leave it alone, I think (but would it surprise users to treat it differently?).
- For HTML output, `<script>` should pretty much always preserve its white space intact. Several tags such as `<pre>`, `<code>`, or `<textarea>` imply an expectation that white space will be preserved, and users will be unpleasantly surprised if they don't. 
- Authors need a way to selectively preserve white space. Some may eschew `<pre>` in favor of something like `<span style="white-space:pre">`. Or in mixed content you get things like `<p>I <em>love</em> <a href="#">Thymeleaf</a>!</p>`.
- What most users will want most of the time for production is to normalize white space from the template except for a few special space-sensitive tags, with as little extra work as possible. For development, though, I can see disabling this so that "view source" is easier to work with. 

Therefore, what I would suggest is:
- Add a bean property for whether white space normalization is enabled by default on all templates. Default to false for backward compatibility. 
- Add a bean property, set-valued, for which elements to preserve white space on by default. Default to `script`, `textarea`, `pre`, `code`. 
- Add an attribute, say `th:space`, to explicitly control white space handling on an element. This would extend to its descendants, unless they override the attribute either explicitly or by being one of the case-sensitive elements. 
- Normalize space, where enabled, in the `normalize-space()` fashion when loading the template. 
- Do not normalize dynamic content such as from `th:text`. Provide a method `#strings.normalizeSpace` in case users wish to do so explicitly.

**Comment by ultraq** [comment]
In Thymeleaf 3, could this be the job of a post-processing dialect developed by someone else, so not part of the Thymeleaf core?  You said something similar in #108, so I'm thinking it'll be the same for here.

**Comment by danielfernandez** [comment]
@ultraq yes exactly, this should do a great external dialect, a part of the ecosystem.

In fact, with Thymeleaf 3.0's architecture I would in fact doubt between implement it as an `IPreProcessor` and an `IPostProcessor`. I think the former (pre-processor) would be more both CPU and memory efficient in most cases.

I'm closing this as declined, the same as #108 was.

**Comment by danielfernandez** [comment]
Actually, better leave this one open, just like #35 and #36, with the `contrib:open` label so that we can use these as a repository of ideas to external contributors.

**Comment by cruftex** [comment]
+1 for keeping it open.

I think Thymeleaf should have a standard solution for this. In what artifact it is included, is another discussion. We might take a look at this and do a contribution, so additional hints and +1 are welcome :)

**Comment by jansen07** [comment]
up. for adding simple/general tags for stripping those extra lines and spaces. please make 
thymeleaf-extras-minify for this issue ..
+1 for #11

**Comment by silkentrance** [comment]
https://stackoverflow.com/questions/46423560/how-do-i-minify-dynamic-html-responses-in-spring

**Comment by GeorgeSalu** [comment]
It's been almost 14 years since this issue was opened on June 11, 2012. It would be great if there was a solution for this within the library itself.
